### PR TITLE
fix(retrieve, pushsync): wait for overdraft peers to refresh

### PR DESCRIPTION
**Pull request by istae**
### Checklist

- [x] I have read the [coding guide](https://github.com/ethersphere/bee/blob/master/CODING.md).
- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [ ] I have filled out the description and linked the related issues.

### CODE REVIEW

File: pkg/skippeers/skippeers_test.go [hunk 1]
```diff
@@ -28,9 +28,12 @@ func TestPeerSkipList(t *testing.T) {
 
 	skipList.Add(addr1, addr3, time.Millisecond*10)
 
-	time.Sleep(time.Millisecond * 11)
+	skipList.PruneExpiresAfter(time.Millisecond)
```
Comment (aloknerurkar) [inline pkg/skippeers/skippeers_test.go:31]: Can we also verify the return value here?

File: pkg/retrieval/retrieval.go [hunk 2]
```diff
@@ -275,10 +304,13 @@ func (s *Service) retrieveChunk(ctx context.Context, peer swarm.Address, done ch
 	// Reserve to see whether we can request the chunk
 	creditAction, err := s.accounting.PrepareCredit(creditCtx, peer, chunkPrice, isOrigin)
 	if err != nil {
+		skip.Add(addr, peer, overDraftRefresh)
 		return
 	}
 	defer creditAction.Cleanup()
 
+	skip.Add(addr, peer, maxDuration)
```
Comment (aloknerurkar) [inline pkg/retrieval/retrieval.go:312]: We add the overdraft peers separately. If there is an error with this request, we will add the peer to the errSkip list in retrieveChunk. If the request was successful we will have the chunk. So I am not sure if there is any point in adding the peer to the local skiplist here. We might end up wasting some CPU to check a list with duplicate peers in the closestPeer function.

Comment (istae) [reply pkg/retrieval/retrieval.go:312]: if we add the peer to the skiplist at the end of the request, a preemtive attempt could try the same peer before the first request returns.

**Review by aloknerurkar — APPROVED** [review]

File: pkg/skippeers/skippeers.go [hunk 3]
```diff
@@ -70,4 +72,15 @@ func (l *List) PruneExpired() {
 			delete(l.skip, k)
 		}
 	}
+
+	return count
+}
+
+func (l *List) Reset() {
```
Comment (istae) [inline pkg/skippeers/skippeers.go:79]: @aloknerurkar @vladopajic @mrekucci just to confirm, do we need this Reset to prevent memory leak?

Comment (aloknerurkar) [reply pkg/skippeers/skippeers.go:79]: I dont think we need it as this skip list is local to the function and once the function returns it should be GCd.

Comment (istae) [reply pkg/skippeers/skippeers.go:79]: i was not sure since it's a map within a map

File: pkg/pushsync/pushsync.go [hunk 4]
```diff
@@ -41,10 +41,12 @@ const (
...
-	defaultTTL         = 30 * time.Second // request time to live
-	preemptiveInterval = 5 * time.Second  // P90 request time to live
-	sanctionWait       = 5 * time.Minute
-	replicationTTL     = 5 * time.Second // time to live for neighborhood replication
+	defaultTTL                       = 30 * time.Second // request time to live
+	preemptiveInterval               = 5 * time.Second  // P90 request time to live
+	sanctionWait                     = 5 * time.Minute
+	replicationTTL                   = 5 * time.Second // time to live for neighborhood replication
+	overDraftRefresh                 = time.Second
+	maxDuration        time.Duration = 1<<63 - 1
```
Comment (mrekucci) [inline pkg/pushsync/pushsync.go:49]: Use `math.MaxInt64` instead.

File: pkg/pushsync/pushsync.go [hunk 5]
```diff
@@ -397,13 +400,40 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
...
 
 			peer, err := nextPeer()
+
+			// no peers left
+			if errors.Is(err, topology.ErrNotFound) {
+				if skip.PruneExpiresAfter(overDraftRefresh) == 0 { //no overdraft peers, we have depleted ALL peers
+					if inflight == 0 {
+						ps.logger.Debug("no peers left", "chunk_address", ch.Address(), "error", err)
+						return nil, err
+					} else {
```
Comment (mrekucci) [inline pkg/pushsync/pushsync.go:410]: Because of the `return` the `else` is unnecessary here.

File: pkg/pushsync/pushsync.go [hunk 6]
```diff
@@ -397,13 +400,40 @@ func (ps *PushSync) pushToClosest(ctx context.Context, ch swarm.Chunk, origin bo
...
+
+			if errors.Is(err, topology.ErrWantSelf) {
+				return nil, err
+			}
+
 			if err != nil {
-				if inflight == 0 {
+				if inflight > 0 {
+					continue
+				} else {
```
Comment (mrekucci) [inline pkg/pushsync/pushsync.go:433]: Because of the `continue` above `else` is unnecessary here.

File: pkg/retrieval/retrieval.go [hunk 7]
```diff
@@ -106,17 +106,20 @@ func (s *Service) Protocol() p2p.ProtocolSpec {
...
 const (
-	retrieveChunkTimeout = 10 * time.Second
+	retrieveChunkTimeout = time.Second * 10
 	preemptiveInterval   = time.Second
+	overDraftRefresh     = time.Second
 	skiplistDur          = time.Minute
 	maxRetrievedErrors   = 32
 	originSuffix         = "_origin"
+
+	maxDuration time.Duration = 1<<63 - 1
```
Comment (mrekucci) [inline pkg/retrieval/retrieval.go:116]: The same as above, use `math.MaxInt64` instead.

File: pkg/retrieval/retrieval.go [hunk 8]
```diff
@@ -220,16 +213,43 @@ func (s *Service) RetrieveChunk(ctx context.Context, chunkAddr, sourcePeerAddr s
...
+					continue
+				}
+
+				// no peers left
+				if errors.Is(res.err, topology.ErrNotFound) {
+					if skip.PruneExpiresAfter(overDraftRefresh) == 0 { //no overdraft peers, we have depleted ALL peers
+						if inflight == 0 {
+							loggerV1.Debug("no peers left", "chunk_address", chunkAddr, "error", res.err)
+							return nil, res.err
+						} else {
```
Comment (mrekucci) [inline pkg/retrieval/retrieval.go:228]: As above, unnecessary `else` statement.

**Review by mrekucci — APPROVED** [review]

**Pull request merged**